group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -983,27 +1052,125 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         return new CoordinatorResult<>(records, response);
     }
 
+    private void replaceStaticMemberInConsumerGroup(
+        String groupId,
+        List<Record> records,
+        ConsumerGroup group,
+        ConsumerGroupMember member,
+        ConsumerGroupMember existingStaticMember,
+        ConsumerGroupMember updatedMember
+    ) {
+        // Write tombstones for the departed static member. Follow the same order as the one followed when fencing
+        // a member
+        records.add(newCurrentAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newTargetAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newMemberSubscriptionTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        // Cancel all the timers of the departed static member.
+        cancelConsumerGroupSessionTimeout(group.groupId(), existingStaticMember.memberId());
+        cancelConsumerGroupRevocationTimeout(group.groupId(), existingStaticMember.memberId());
+        // Write a record corresponding to the new member
+        records.add(newMemberSubscriptionRecord(groupId, updatedMember));
+        TargetAssignmentBuilder.TargetAssignmentResult assignmentResult = computeTargetAssignment(group, group.groupEpoch(), member, updatedMember);
+        records.addAll(assignmentResult.records());
+        records.add(newCurrentAssignmentRecord(groupId, updatedMember));
+    }
+
+    private TargetAssignmentBuilder.TargetAssignmentResult computeTargetAssignment(
+        ConsumerGroup group,
+        int groupEpoch,
+        ConsumerGroupMember member,
+        ConsumerGroupMember updatedMember) {
+        String preferredServerAssignor = group.computePreferredServerAssignor(
+            member,
+            updatedMember
+        ).orElse(defaultAssignor.name());
+
+        String groupId = group.groupId();
+        Map<String, TopicMetadata> subscriptionMetadata = group.subscriptionMetadata();
+        String memberId = member.memberId();
+        try {
+            TargetAssignmentBuilder.TargetAssignmentResult assignmentResult =
+                new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(preferredServerAssignor))
+                    .withMembers(group.members())
```

Comment:
I think that the old member will be in `members` so the computed target assignment is incorrect. We need to remove it with `removeMember` and we also need to set the target assignment of the new member from the old one.

Reply:
I missed the `removeMembers` part. Thanks for pointing it out. Regarding 

> we also need to set the target assignment of the new member from the old one.

I assumed, this [block of code](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java#L262-L288)  should take care of it. Was my assumption wrong?

Reply:
Well, at the moment, a new target assignment will be computed for the new static member and that block of code will indeed create the record for it. What I meant is that the new static member should actually reuse the target assignment of the previous member (vs computing a new one). Then we only need to recompute the assignment if there is a group epoch bump.

Reply:
That makes sense and thanks for the explanation. I have now changed the code to remove the existing static member and add the new static member. Rest of the state would remain as is.

Reply:
Also, regarding 

> Then we only need to recompute the assignment if there is a group epoch bump.

Given that now we will have a group epoch bump whenever a static member re-joins with a different subscription, should this be mentioned in the KIP? As we noticed, this is a deviation from how the static member rejoining with a different subscription case as of today. WDYT?